dpnp/dpnp_utils/dpnp_utils_linearalgebra.py

```diff
@@ -2007,61 +2128,56 @@ def dpnp_matmul(
             dep_events_list,
             host_tasks_list,
             copy_flag=not contig_flag,
-            dtype=gemm_dtype,
+            dtype=compute_dtype,
         )
         contig_flag = _define_contig_flag(x2)
         x2 = _copy_array(
             x2,
             dep_events_list,
             host_tasks_list,
             copy_flag=not contig_flag,
-            dtype=gemm_dtype,
+            dtype=compute_dtype,
         )
 
-        # TODO: investigate usage of gemv (gemv_batch) function
-        # from BLAS when one of the inputs is a vector to
-        # gain performance.
-        # TODO: investigate usage of syrk function from BLAS in
-        # case of a.T @ a and a @ a.T to gain performance.
         row_major = True
-        if x1_is_2D and x2_is_2D:
+        ht_blas_ev = []
+        if call_flag == "gemm":
             ht_blas_ev, _, row_major = bi._gemm(
                 exec_q,
                 dpnp.get_usm_ndarray(x1),
                 dpnp.get_usm_ndarray(x2),
-                dpnp.get_usm_ndarray(result),
+                dpnp.get_usm_ndarray(res),
                 dep_events_list,
             )
-            host_tasks_list.append(ht_blas_ev)
-        else:
-            result = _gemm_batch_matmul(
+        elif call_flag == "gemm_batch":
+            res = _gemm_batch_matmul(
                 exec_q,
                 x1,
                 x2,
-                result,
+                res,
                 dep_events_list,
             )
 
+        if ht_blas_ev:
+            host_tasks_list.append(ht_blas_ev)
         dpctl.SyclEvent.wait_for(host_tasks_list)
         if not row_major:
             # TODO: investigate the possibility of defining result
             # array with "F" order for this case
-            result = dpnp.ascontiguousarray(
-                dpnp.reshape(result.ravel(), result.shape, order="F")
+            res = dpnp.ascontiguousarray(
```

Comment:
we have exactly the same code inside `_gemm_batch_matmul`, so we shouldn't run it twice for `call_flag == "gemm_batch"`